### ISSUE TITLE
🐞 SuperADD: last training batch is stored unsynchronised (`non_blocking` MPS→CPU copy), corrupting the memory bank outside `Engine.fit`

### ISSUE THREAD

**Issue by tkafka** [open-edge-platform/anomalib#3810]
### Describe the bug

Hello, this is an issue discovered by Claude, seems to point to the actual problem:

### Describe the bug

`SuperADDModel.forward` stores the training embeddings with a non-blocking device→host copy
([`src/anomalib/models/image/super_add/torch_model.py`, line 415 on `main`](https://github.com/open-edge-platform/anomalib/blob/main/src/anomalib/models/image/super_add/torch_model.py#L415)):

```python
self.embedding_store[layer_idx].append(embedding_reshaped.to("cpu", non_blocking=True))
```

Nothing synchronises the device before `subsample_embedding()` reads `embedding_store`. A non-blocking
device→host copy may not have landed when the CPU reads it
([PyTorch guide on `non_blocking`](https://docs.pytorch.org/tutorials/intermediate/pinmem_nonblock.html)).

On Apple Silicon (MPS), when the model is fed without Lightning in between, the **last batch** in the store is
wrong every time. In the repro below, exactly the last 4 of 16 images (3136 of 12544 rows) differ from a
synchronous recomputation, by up to 328. Depending on what the uninitialised memory holds, the memory bank
built from that store contains NaN or silently wrong features. In our pipeline a `vit_small_patch16_dinov3`
bank had 518 NaN values and every anomaly map became NaN. With `vit_base_patch16_dinov3` the bank was finite but
built from wrong data. No error is raised in either case.

With `torch.mps.synchronize()` before reading the store, all rows match exactly.

Scope: through the documented `Engine.fit` path we could **not** reproduce it; the bank matched a synchronous
recomputation to float noise. Lightning apparently synchronises after the last `training_step`, for example
through the dummy loss. The bug affects code that drives the torch model directly, such as a custom teach-in
loop that builds a bank on the fly. Correctness also depends on an implicit synchronisation elsewhere, which
could break with other Lightning versions or loops. We did not test CUDA. The PyTorch guide above says a
non-blocking device→host copy needs a synchronisation there too.

Suggested fix: drop `non_blocking=True` (one extra synchronisation per batch during a single-epoch feature
pass is negligible). Alternatively, synchronise the compute device at the start of `subsample_embedding()`:

```python
if device.type == "cuda":
    torch.cuda.synchronize(device)
elif device.type == "mps":
    torch.mps.synchronize()
```

### Model

Other: SuperADD

### Dataset

N/A

### Model

Other (please specify in the field below)

### Steps to reproduce the behavior

On an Apple Silicon Mac:

```python
import torch
from anomalib.models import SuperADD

dev = torch.device("mps")
m = SuperADD(backbone="vit_small_patch16_dinov3", layers=[2, 5, 8, 11], patch_size=448, patch_overlap=64,
             post_processor=False, evaluator=False, visualizer=False)
tm = m.model.to(dev).train()
batches = torch.rand(16, 3, 448, 448, generator=torch.Generator().manual_seed(0)).split(4)
for b in batches:
    tm(b.to(dev))  # training branch: embedding.to("cpu", non_blocking=True)
# torch.mps.synchronize()  # uncommenting this line makes every row match
stored = torch.cat(tm.embedding_store[3])
with torch.no_grad():
    ref = torch.cat([tm.patch_exec(b.to(dev))[3].reshape(-1, stored.shape[-1]).cpu() for b in batches])
bad = ((stored - ref).abs() > 1e-3).any(1)
print(f"{int(bad.sum())} of {len(stored)} stored rows differ from a synchronous recomputation")
m.fit()
print("memory bank finite:", bool(torch.isfinite(tm.memory_bank).all()))
```

Output:

```
3136 of 12544 stored rows differ from a synchronous recomputation
memory bank finite: True
```

With the `torch.mps.synchronize()` line uncommented:

```
0 of 12544 stored rows differ from a synchronous recomputation
memory bank finite: True
```

3136 rows = 4 images × 28 × 28 tokens, i.e. exactly the last batch. With real images the corrupted rows
sometimes contain NaN (518 values in a ViT-S bank in our case), and then every anomaly map is NaN.

### OS information

- OS: macOS 27.0 (26A428)
- Hardware: Apple M3 Max, MPS backend
- Python: 3.12.9
- torch 2.14.0, lightning 2.6.6, timm 1.0.30

### Expected behavior

The embeddings in `embedding_store`, and the memory bank built from them, equal the features computed by
the backbone, independent of device timing. This should hold whether the model is trained through
`Engine.fit` or called directly.

### Pip/GitHub

pip

### What version/branch did you use?

anomalib 2.6.2

### Configuration YAML

```yaml
not used; the model is constructed in Python as in the repro above
```

### Logs

```shell
no error or warning is emitted; the corruption is silent
```

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct
